### DISCUSSION TITLE
Welcome to Mapzy Discussions!

### DISCUSSION THREAD

**Discussion by shafy** [mapzy/mapzy#61] [Announcements]
## 👋 Welcome!
Discussions is the best place to:
- Learn about Mapzy
- Ask questions and get help about Mapzy's functionalities
- Share ideas for new features and feedback
- Get help if you're self-hosting Mapzy
- Have fun

Welcome, and looking forward to your participation!

**Comment by thebah19** [comment]
Hello,

Can Mapzy be hosted on bluehost.com?

**Reply by shafy** [reply]
Hey @thebah19

I don't have any experience with bluehost, but it looks like their offering is geared more towards hosting WordPress or "conventional" websites. In general, if you can run a Docker container on bluehost you can run Mapzy. Although it looks like it might be possible to do that (see [here](https://stackoverflow.com/questions/46509830/can-i-install-docker-over-cpanel)), it's definitely not the most straight-forward way.

The simplest way to self-host Mapzy without any knowledge of running Docker is to use [Elestio](https://elest.io/). There, you can just create an account and choose [Mapzy as an app to launch](https://elest.io/open-source/mapzy).

Is there a reason you need to use bluehost strictly?

And, of course, if self-hosting is not a requirement, you can always use our [hosted service](https://mapzy.io).

**Comment by Pranav4399** [comment]
I'm trying to set up mapzy locally. I'm successful with running the dockerfile.dev image. Can you help me with the .env.example file? What details should be filled there? If possible, can someone share their local .env.example(Without any credentials though) for reference? 

Also, I tried to run this without executing  these two commands, 

` Set the .env files and fill in all env vars`
`cp .env.example .env`

`# Create and migrate the database`
`rails db:create db:migrate`

and received an error saying rails server couldn't start. 
![image](https://user-images.githubusercontent.com/28348429/188699429-8bc19376-21c0-447e-9f92-d901443cdbc2.png)

Any help : ) ?

**Reply by shafy** [reply]
Hey there!

For the `.env` file, you can just create `.env` file in the root folder of project (`cp .env.example .env` just means "copy `.env.example` to a new file called `.env`). As for the environment variables, assuming you have a Postgres server and Redis server running separately and just want to star the Dockerfile, you'll need to add these in `.env`

```
RAILS_ENV=development
ALLOW_REGISTRATION=true

# if you're running Postgres and Redis, use the values from there in the following env vars
DATABASE_NAME=
DATABASE_USER=
DATABASE_PASSWORD=
REDIS_PORT=

# Create a mapbox.com account and retrieve your access token
MAPBOX_ACCESS_TOKEN=
```

Alternatively, if you don't want to run your own Postgres and Redis servers, you can use docker compose. Then you need can just set the values you want for `DATABASE_NAME`, `DATABASE_USER` and `DATABASE_PASSWORD` and remove `REDIS_PORT`. 

You can then start everything up with `docker-compose -f docker-compose.dev.yml up`.

As a third alternative, you can open up this repo in Gitpod, which is a cloud-based dev environment that's preconfigured (well, you need to set the env variables `MAPBOX_ACCESS_TOKEN` and `ALLOW_REGISTRATION` in Gitpod). To open it in Gitpod, just follow [this link](https://gitpod.io/#https://github.com/mapzy/mapzy). Once it has loaded, just run `bin/dev` in the console to start the Rails server.

There's also some more info in our [self-hosting doc](https://github.com/mapzy/mapzy/blob/main/SELFHOSTING.md) (probably already seen it).

Let me know if this works or what erros you run into. It can take a while if you're not used to running Rails applications.

**Comment by Pranav4399** [comment]
Hello. I followed the steps mentioned above and I progressed a lot with the setup. However, I'm facing some issues with a command in the file `20220617113632_create_api_keys.rb` as shown in the image. 

I'm new to the entire rails applications. So, any help with this? Also, can you suggest some good resources which will help me to understand rails which will eventually help me understand mapzy too

![image](https://user-images.githubusercontent.com/28348429/189175478-82393e84-0ba4-4a7b-8ff1-966d4bf72f61.png)

**Reply by shafy** [reply]
Hi! Can you describe the steps to reproduce this?

Regarding resources, I think the official [Ruby on Rails Guide](https://guides.rubyonrails.org/) is great place to start. If you want to dive deeper after that, I suggest to build your own app from scratch that does something simple. There are a lot of good tutorials if you search (I can't recommend a particular one).

**Comment by Pranav4399** [comment]
Hello ! So, I followed the steps mentioned in your previous comment to setup the local development codebase. I followed the steps including `docker-compose -f docker-compose.dev.yml up`. It automatically setup the postgres, redis and web app containers on my local.

![image](https://user-images.githubusercontent.com/28348429/189706092-1a3a1a02-bbf8-44b4-aee4-c9316adfe89b.png)

I faced a different error at this stage which I resolved by removing line 20 from the docker-compose.dev.yml file - `entrypoint: ["sh", "/app/docker-entrypoint.sh"]` and was successful in accessing the rails web page in localhost:3000 port. I then ended up in this page - 

![image](https://user-images.githubusercontent.com/28348429/189706580-995563fd-221b-4e24-9f67-5e89c2677a0e.png)

On clicking Run pending migrations, I get the afore mentioed error - 

![image](https://user-images.githubusercontent.com/28348429/189706759-6f7ac99f-0e44-41b5-bf5d-762acd744331.png)

This line exists in the file `mapzy\db\migrate\20220617113632_create_api_keys.rb` which I believe is auto generated from the DB(Correct me if I'm wrong). This is something to do with running tasks using Rake but I'm unable to figure it out.

**Reply by shafy** [reply]
Hey thanks for the detailed reproduction steps. You shouldn't need to remove `entrypoint: ["sh", "/app/docker-entrypoint.sh"]` - in this file, the migrations are run and that's why you get the later error that there are migrations pending. In this setup, running migrations by pressing the "Run pending migrations" button doesn't work.

Can you tell me what error you were seeing when you leave the line `entrypoint: ["sh", "/app/docker-entrypoint.sh"]` in? Let's solve this, and then everything else should work.

> This line exists in the file mapzy\db\migrate\20220617113632_create_api_keys.rb which I believe is auto generated from the DB(Correct me if I'm wrong). This is something to do with running tasks using Rake but I'm unable to figure it out.

No, it's the other way around. The database schema is created according to these migration files :-)

**Reply by Pranav4399** [reply]
Hello. Sorry for the long delay in response. I kept this line `entrypoint: ["sh", "/app/docker-entrypoint.sh"]` in the docker-compose.dev.yml file and ran the below command. This is the response that I got. I was able to get the redis and postgres instance up and running but the website instance failed

```
mapzy>docker-compose -f docker-compose.dev.yml up
[+] Running 3/3
 - Container mapzy-redis-1  Created                                                                                0.0s
 - Container mapzy-db-1     Created                                                                                0.0s
 - Container mapzy-web-1    Recreated                                                                              0.2s
Attaching to mapzy-db-1, mapzy-redis-1, mapzy-web-1
mapzy-redis-1  | 1:C 17 Sep 2022 18:17:46.114 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
mapzy-redis-1  | 1:C 17 Sep 2022 18:17:46.114 # Redis version=7.0.4, bits=64, commit=00000000, modified=0, pid=1, just started
mapzy-redis-1  | 1:C 17 Sep 2022 18:17:46.114 # Warning: no config file specified, using the default config. In order to specify a config file use redis-server /path/to/redis.conf
mapzy-redis-1  | 1:M 17 Sep 2022 18:17:46.115 * monotonic clock: POSIX clock_gettime
mapzy-redis-1  | 1:M 17 Sep 2022 18:17:46.117 * Running mode=standalone, port=6379.
mapzy-redis-1  | 1:M 17 Sep 2022 18:17:46.117 # Server initialized
mapzy-redis-1  | 1:M 17 Sep 2022 18:17:46.118 # WARNING overcommit_memory is set to 0! Background save may fail under low memory condition. To fix this issue add 'vm.overcommit_memory = 1' to /etc/sysctl.conf and then reboot or run the command 'sysctl vm.overcommit_memory=1' for this to take effect.
mapzy-redis-1  | 1:M 17 Sep 2022 18:17:46.120 * Loading RDB produced by version 7.0.4
mapzy-redis-1  | 1:M 17 Sep 2022 18:17:46.120 * RDB age 323 seconds
mapzy-redis-1  | 1:M 17 Sep 2022 18:17:46.120 * RDB memory usage when created 1.53 Mb
mapzy-redis-1  | 1:M 17 Sep 2022 18:17:46.120 * Done loading RDB, keys loaded: 11, keys expired: 1.
mapzy-redis-1  | 1:M 17 Sep 2022 18:17:46.120 * DB loaded from disk: 0.000 seconds
mapzy-redis-1  | 1:M 17 Sep 2022 18:17:46.120 * Ready to accept connections
mapzy-db-1     |
mapzy-db-1     | PostgreSQL Database directory appears to contain a database; Skipping initialization
mapzy-db-1     |
mapzy-db-1     | 2022-09-17 18:17:47.916 UTC [1] LOG:  starting PostgreSQL 14.5 (Debian 14.5-1.pgdg110+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 10.2.1-6) 10.2.1 20210110, 64-bit
mapzy-db-1     | 2022-09-17 18:17:47.997 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
mapzy-db-1     | 2022-09-17 18:17:48.004 UTC [1] LOG:  listening on IPv6 address "::", port 5432
mapzy-db-1     | 2022-09-17 18:17:48.029 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
mapzy-web-1    | /app/docker-entrypoint.sh: 2: set: Illegal option -
mapzy-db-1     | 2022-09-17 18:17:48.088 UTC [26] LOG:  database system was shut down at 2022-09-17 18:12:32 UTC
mapzy-db-1     | 2022-09-17 18:17:48.148 UTC [1] LOG:  database system is ready to accept connections
mapzy-web-1 exited with code 2
mapzy-web-1 exited with code 2
mapzy-web-1    | /app/docker-entrypoint.sh: 2: set: Illegal option -
mapzy-web-1 exited with code 2
```

I remember seeing this and removing that line that got the web running but in turn faced the migrations issue. Looking forward to hearing from you.

**Reply by shafy** [reply]
Thanks @Pranav4399! On what operatign system are you? Windows?
Can you try to change the `"sh"` to `"bash"` on line 20 in the `docker-compose.dev.yml` file? Like this:
```
entrypoint: ["bash", "/app/docker-entrypoint.sh"]
```

And then try again? You might need to delete/flush the docker images and containers before.
